modules/subtitle-timeshift-workflowoperation/src/main/java/org/opencastproject/workflow/handler/subtitletimeshift/SubtitleTimeshiftWorkflowOperationHandler.java

```diff
@@ -0,0 +1,291 @@
+  private static final String VIDEO_SOURCE_FLAVOR_CFG_KEY = "video-source-flavor";
+  private static final String TARGET_FLAVOR_CFG_KEY = "target-flavor";
+
+  /** The workspace collection name */
+  private static final String COLLECTION = "subtitles";
+
+  /** The logging facility */
+  private static final Logger logger = LoggerFactory.getLogger(SubtitleTimeshiftWorkflowOperationHandler.class);
+
+  /**
+   * Reference to the workspace service
+   */
+  private Workspace workspace = null;
+
+
+  /**
+   * OSGi setter for the workspace class
+   *
+   * @param workspace an instance of the workspace
+   */
+  @Reference
+  public void setWorkspace(Workspace workspace) {
+    this.workspace = workspace;
+  }
+
+
+  @Override
+  public WorkflowOperationResult start(WorkflowInstance workflowInstance, JobContext context)
+          throws WorkflowOperationException {
+
+    MediaPackage mediaPackage = workflowInstance.getMediaPackage();
+    logger.info("Starting subtitle timeshift workflow for mediapackage: {}", mediaPackage.getIdentifier().toString());
+
+    // get flavor from workflow configuration
+    final WorkflowOperationInstance operation = workflowInstance.getCurrentOperation();
+    MediaPackageElementFlavor configuredSubtitleFlavor;
+    MediaPackageElementFlavor configuredVideoFlavor;
+    MediaPackageElementFlavor configuredTargetFlavor;
+    try {
+      configuredSubtitleFlavor = MediaPackageElementFlavor.parseFlavor(
+          Objects.toString(operation.getConfiguration(SUBTITLE_SOURCE_FLAVOR_CFG_KEY)));
+      configuredVideoFlavor = MediaPackageElementFlavor.parseFlavor(
+          operation.getConfiguration(VIDEO_SOURCE_FLAVOR_CFG_KEY));
+      configuredTargetFlavor = MediaPackageElementFlavor.parseFlavor(
+          operation.getConfiguration(TARGET_FLAVOR_CFG_KEY));
+    } catch (Exception e) {
+      throw new WorkflowOperationException("Couldn't parse subtitle-timeshift workflow configurations.", e);
+    }
+
+    // In this block we try to get the subtitle and video track for this workflow
+    Track originalSubtitleTrack;
+    Track videoTrack;
+    try {
+      // Get the subtitles and videos from the mediapackage
+      Track[] subtitleTracks = mediaPackage.getTracks(configuredSubtitleFlavor);
+      Track[] videoTracks = mediaPackage.getTracks(configuredVideoFlavor);
+      // Check if we found exactly 1 subtitle track and 1 video track
+      boolean foundTracksAreValid = validateTracks(configuredSubtitleFlavor.toString(), subtitleTracks.length,
+          configuredVideoFlavor.toString(), videoTracks.length);
+      // If we found the wrong amount, we skip the workflow operation
+      if (!foundTracksAreValid) {
+        logger.info("Skipping subtitle-timeshift workflow, because it couldn't find "
+            + "the right amount of subtitle and video tracks. Mediapackage: {}", mediaPackage.getIdentifier());
+        return createResult(mediaPackage, Action.SKIP);
```

Comment:
Thank you for the review! I am currently working on the code.

This part you mentioning here is not for checking the amount of arguments, rather it counts how many tracks were found with the configured flavor. If someone configures `*/source` as the `video-source-flavor` for example, the operation will find 2 tracks (in the case of a dual stream) and the operation expects 1 video. It would be no Problem to throw an error at this point. What would you suggest for the following cases?:

1. The operation finds no video with the given `video-source-flavor`: throw an error?
2. The operation finds more than 1 video: throw an error?
Doing it like this would be okay for me. So what we saying then is: "if the operation doesn't find exactly 1 video with the configured flavor, the user did not configured the workflow operation correctly and should fix it".

I am also changing the behavior of what the operations does, when it finds a certain amount of subtitle files:

1. The operation finds no subtitle file: simply skip operation?
2. The operation finds >= 1 subtitle files: timeshift all of them?
Currently the operation will skip if it finds more than 1 subtitle file, but you suggested supporting the shifting of multiple files at once (this would be the case for example when there are more languages involved). I think you are right and I will implement it this way.

Reply:
That makes sense to me.